drivers/sensor/apds9960/apds9960.c

```diff
@@ -0,0 +1,455 @@
+/* sensor_apds9960.c - driver for APDS9960 ALS/RGB/gesture/proximity sensor */
+
+/*
+ *
+ *Copyright (c) 2017 Intel Corporation
+ *
+ *SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include "apds9960.h"
+
+static int apds9960_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct apds9960_data *data = dev->driver_data;
+	u8_t cmsb, clsb, rmsb, rlsb, blsb, bmsb, glsb, gmsb;
+	u8_t pdata;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	/* Read lower byte following by MSB  Ref: Datasheet : RGBC register */
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_CDATAL_REG, &clsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_CDATAH_REG, &cmsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_RDATAL_REG, &rlsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_RDATAH_REG, &rmsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_GDATAL_REG, &glsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_GDATAH_REG, &gmsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_BDATAL_REG, &blsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_BDATAH_REG, &bmsb) < 0) {
+		return -EIO;
+	}
+
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_PDATA_REG, &pdata) < 0) {
+		return -EIO;
+	}
+
+	data->sample_c = (cmsb << 8) + clsb;
+	data->sample_r = (rmsb << 8) + rlsb;
+	data->sample_g = (gmsb << 8) + glsb;
+	data->sample_b = (bmsb << 8) + blsb;
+	data->pdata    = pdata;
+
+	return 0;
+}
+
+static int apds9960_channel_get(struct device *dev,
+				enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct apds9960_data *data = dev->driver_data;
+
+	switch (chan) {
+	case SENSOR_CHAN_LIGHT:
+		val->val1 = data->sample_c;
+		val->val2 = 0;
+		break;
+	case (SENSOR_CHAN_RED):
+		val->val1 = data->sample_r;
+		val->val2 = 0;
+		break;
+	case (SENSOR_CHAN_GREEN):
+		val->val1 = data->sample_g;
+		val->val2 = 0;
+		break;
+	case (SENSOR_CHAN_BLUE):
+		val->val1 = data->sample_b;
+		val->val2 = 0;
+		break;
+	case (SENSOR_CHAN_PROX):
+		val->val1 = data->pdata;
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static bool apds9960_setproxintlowthresh(struct device *dev, u8_t threshold)
+{
+	struct apds9960_data *data = dev->driver_data;
+
+	if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_PILT_REG, threshold) < 0) {
+		SYS_LOG_DBG(" Failed");
+		return false;
+	}
+
+	return true;
+}
+
+static bool apds9960_setproxinthighthresh(struct device *dev, u8_t threshold)
```

Comment:
This function and all of its friends squash the errno returns codes they get to true/false and the caller then converts false to EIO. Instead these functions should directly return the errno codes they catch rather than squashing to true/false, and the caller should likewise propagate the errno.